pkg/cosign/cosign.go

```diff
@@ -50,37 +51,39 @@ func Initialize(client kubernetes.Interface, namespace, serviceAccount string, i
 }
 
 // VerifySignature verifies that the image has the expected key
-func VerifySignature(imageRef string, key []byte, repository string, log logr.Logger) (digest string, err error) {
-	pubKey, err := decodePEM(key)
+func VerifySignature(imageRef string, key string, repository string, log logr.Logger) (digest string, err error) {
+	ctx := context.Background()
+
+	var pubKey signature.Verifier
+
+	if strings.HasPrefix(key, "-----BEGIN PUBLIC KEY-----") {
+		pubKey, err = decodePEM([]byte(key))
+	} else {
+		pubKey, err = sigs.PublicKeyFromKeyRef(ctx, key)
+		if err != nil {
+			return "", errors.Wrap(err, "loading public key")
+		}
+	}
+
+	os.Setenv("COSIGN_REPOSITORY", repository)
```

Comment:
@developer-guy - that would work, but allow only a single COSIGN_REPOSITORY for the Kyverno deployment. The prior implementation allowed specifying alternate signature repositories per matching image in a policy rule, which was the requirement. For example:

```yaml
      - image: "ghcr.io/kyverno/*"
        repository: ghcr.io/kyverno/signatures
        key: |-
          -----BEGIN PUBLIC KEY-----
          MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE8nXRh950IZbRj8Ra/N9sbqOPZrfM
          5/KAQN0/KjHcorm/J5yctVd7iEcnessRQjU917hmKO6JWVGHpDguIyakZA==
          -----END PUBLIC KEY-----          
      - image: "docker.io/kyverno/*"
        repository: docker.io/kyverno/signatures
        key: |-
          -----BEGIN PUBLIC KEY-----
          MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE8nXRh950IZbRj8Ra/N9sbqOPZrfM
          5/KAQN0/KjHcorm/J5yctVd7iEcnessRQjU917hmKO6JWVGHpDguIyakZA==
          -----END PUBLIC KEY-----          

```

Reply:
got it, this commit will fix this IMHO. @JimBugwadia 
https://github.com/kyverno/kyverno/pull/2607/commits/8f95c0f1931c81c812febb69be978276e8dffe0d

Reply:
Hi @developer-guy - since setting COSIGN_REPOSITORY leads to shared state, it can impact common behaviors (across policy runs for different namespaces, etc.), and this could lead to some unexpected behaviors when the policy rules are executed concurrently. 

We could add locking, etc. but that seems unnecessary as we can just set the registry via `cosignOpts.SignatureRepo`. What do you think?

```golang

	cosignOpts.SignatureRepo = ref.Context()
	if repository != "" {
		signatureRepo, err := name.NewRepository(repository)
		if err != nil {
			return "", errors.Wrapf(err, "failed to parse signature repository %s", repository)
		}

		cosignOpts.SignatureRepo = signatureRepo
	}

```

Reply:
hello @JimBugwadia, thanks for the feedback, and you're absolutely right. I did the same logic with different code according to updates on cosign side. Here is the following code I've just added to solve this `shared state` problem caused by using an environment variable:

```shell
var opts []remote.Option
	ro := options.RegistryOptions{}
	opts, err = ro.ClientOpts(ctx)
	if err != nil {
		return "", errors.Wrap(err, "constructing client options")
	}

	if repository != "" {
		signatureRepo, err := name.NewRepository(repository)
		if err != nil {
			return "", errors.Wrapf(err, "failed to parse signature repository %s", repository)
		}
		opts = append(opts, remote.WithTargetRepository(signatureRepo))
	}
```

Reply:
This looks great, @developer-guy!